### PR TITLE
Brent/backend66 review endpoints

### LINKED ISSUE #66: Create GET/POST/PATCH/DELETE endpoints for reviews
## Feature Description
Initial functional creation of all /review endpoints for the listing microservice

## Problem You're Trying to Solve
The front end needs to create, edit and delete reviews

## Related External Resources
[SRS doc section 4.4.1](https://docs.google.com/document/d/1DNqexMBf4MpYPCxWsT3plUDAjDSbdJwUzw-UyZX-vDs/edit)

### PR DESCRIPTION
**Pull request by bmachado18**
# Description
Updated the testing framework to cypress. Added tests for TDD for each CRUD API for reviews. Implemented each endpoint, and ensured passing in CI. Note that the getReview endpoint is not in the API spec, but I included it anyway for full CRUD.

I am just waiting on the OpenAPI spec PR to be approved so I can progress from a draft PR. See https://github.com/matt-lebl/uvic-market-openapi/pull/18.

Closes #66 

## How to Test
```docker-compose down -v```
```docker-compose up --build -d review database datalayer```
```docker exec -it martletplace_review bash```
```npm test```

## Checklist
- [x] The code includes tests if relevant
- [x] I have *actually* self-reviewed my changes and done QA

### CODE REVIEW

File: apps/backend/review/cypress.config.ts [hunk 1]
```diff
@@ -0,0 +1,9 @@
+import { defineConfig } from "cypress";
+
+export default defineConfig({
+  e2e: {
+    baseUrl: 'http://localhost:8213',
+    specPattern: 'cypress/e2e/**/*.cy.{js,jsx,ts,tsx}',
```
Comment (MNThomson) [inline apps/backend/review/cypress.config.ts:6]: These aren't e2e tests

File: apps/backend/review/cypress/e2e/createReviewTest.cy.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
+describe('Create Reviews Endpoint', () => {
+  const baseUrl = 'http://localhost:8213/api/review';
```
Comment (MNThomson) [inline apps/backend/review/cypress/e2e/createReviewTest.cy.ts:2]: I don't like hardcoding this but not sure there's a better way

File: apps/backend/review/src/createReview.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+  db: IDatabase<object>,
+) => {
+  // TODO: AUTHENTICATION
+  const userID = 1; // placeholder for authenticated user ID
+  const { stars, comment, listingID } = req.body;
+
+  if (!stars || !comment || !listingID) {
+    console.error("missing parameter in request");
+    return res.status(400).json({ error: "missing parameter in request" });
+  }
```
Comment (MNThomson) [inline apps/backend/review/src/createReview.ts:17]: Do we need any additional validation here? E.g. stars cannot be > 5, etc?

File: apps/backend/review/src/createReview.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+  try {
+    const createdReview = await db.one(
+      `INSERT INTO reviews (listing_id, user_id, review, rating_value, created_at, modified_at)
+       VALUES ($1, $2, $3, $4, NOW(), NOW())
```
Comment (MNThomson) [inline apps/backend/review/src/createReview.ts:22]: Database handles these
```suggestion
      `INSERT INTO reviews (listing_id, user_id, review, rating_value)
       VALUES ($1, $2, $3, $4)
```

File: apps/backend/review/src/createReview.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+       RETURNING *`,
```
Comment (MNThomson) [inline apps/backend/review/src/createReview.ts:23]: Let's return only what we need

File: apps/backend/review/src/createReview.ts [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+
+    const responseReview = {
+      review_id: createdReview.review_id,
+      reviewerName: reviewerProfile.name,
+      stars: createdReview.rating_value,
+      comment: createdReview.review,
+      userID: reviewerProfile.userID,
+      listingID: createdReview.listing_id,
+      dateCreated: createdReview.created_at,
+      dateModified: createdReview.modified_at,
```
Comment (MNThomson) [inline apps/backend/review/src/createReview.ts:42]: Are these in the api spec?

File: apps/backend/review/src/getReview.ts [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    const reviewQuery = `
+      SELECT 
+        r.review_id AS "review_id",
+        ru.username AS "reviewerName",
+        r.rating_value AS "stars",
+        r.review AS "comment",
+        r.user_id AS "userID",
+        r.listing_id AS "listingID",
+        r.created_at AS "dateCreated",
+        r.modified_at AS "dateModified"
```
Comment (MNThomson) [inline apps/backend/review/src/getReview.ts:28]: Are these in the api spec?

File: apps/backend/review/src/index.ts [hunk 8]
```diff
@@ -1,25 +1,36 @@
...
 
 const PORT = 8213;
 
 const app = express();
+
 app.use(morgan("dev"));
 app.use(express.json());
 
+const pgp = pgPromise();
+const DB_ENDPOINT = process.env.DB_ENDPOINT || "";
```
Comment (MNThomson) [inline apps/backend/review/src/index.ts:17]: This should error, in no scenario do we want it to try and connect to a database at empty string

File: apps/backend/review/src/updateReview.ts [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+  db: IDatabase<object>,
+) => {
+  //TODO: AUTHENTICATION
+  const { id } = req.params;
+  const { stars, comment, listingID } = req.body;
+
+  if (!stars || !comment || !listingID) {
+    console.error("missing parameter in request");
+    return res.status(400).json({ error: "missing parameter in request" });
+  }
```
Comment (MNThomson) [inline apps/backend/review/src/updateReview.ts:17]: Any validation?

File: apps/backend/review/src/updateReview.ts [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+
+  try {
+    const updatedReview = await db.oneOrNone(
+      `UPDATE reviews
+       SET review = $1,
+           rating_value = $2,
+           modified_at = NOW(),
```
Comment (MNThomson) [inline apps/backend/review/src/updateReview.ts:24]:
```suggestion
```

File: apps/backend/review/src/updateReview.ts [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+            listing_id = $3
+       WHERE review_id = $4
+       RETURNING *`,
```
Comment (MNThomson) [inline apps/backend/review/src/updateReview.ts:27]: Return specific things

File: apps/backend/review/src/updateReview.ts [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+      console.error("review not found");
+      return res.status(404).json({ error: "Review not found" });
+    }
+
+    const responseReview = {
+      review_id: updatedReview.review_id,
+      stars: updatedReview.rating_value,
+      comment: updatedReview.review,
+      listingID: updatedReview.listing_id,
+      dateModified: updatedReview.modified_at,
```
Comment (MNThomson) [inline apps/backend/review/src/updateReview.ts:41]: Is this in api spec?

**Review by MNThomson — CHANGES_REQUESTED** [review]

File: apps/backend/review/src/createReview.ts (hunk 3 above)
Comment (bmachado18) [reply apps/backend/review/src/createReview.ts:17]: Nope. Swaggerhub has no error defined for invalid stars, and is impossible to click over 5 stars

File: apps/backend/review/src/index.ts (hunk 8 above)
Comment (bmachado18) [reply apps/backend/review/src/index.ts:17]: linter isn't happy with this. [nodemon] restarting due to changes...
[nodemon] starting `ts-node src/index.ts`
/app/service/node_modules/ts-node/src/index.ts:859
    return new TSError(diagnosticText, diagnosticCodes, diagnostics);
           ^
TSError: ⨯ Unable to compile TypeScript:
src/index.ts(19,16): error TS2345: Argument of type 'string | undefined' is not assignable to parameter of type 'string | IConnectionParameters<IClient>'.
  Type 'undefined' is not assignable to type 'string | IConnectionParameters<IClient>'.

    at createTSError (/app/service/node_modules/ts-node/src/index.ts:859:12)
    at reportTSError (/app/service/node_modules/ts-node/src/index.ts:863:19)
    at getOutput (/app/service/node_modules/ts-node/src/index.ts:1077:36)
    at Object.compile (/app/service/node_modules/ts-node/src/index.ts:1433:41)
    at Module.m._compile (/app/service/node_modules/ts-node/src/index.ts:1617:30)
    at Module._extensions..js (node:internal/modules/cjs/loader:1416:10)
    at Object.require.extensions.<computed> [as .ts] (/app/service/node_modules/ts-node/src/index.ts:1621:12)
    at Module.load (node:internal/modules/cjs/loader:1208:32)
    at Function.Module._load (node:internal/modules/cjs/loader:1024:12)
    at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:174:12) {
  diagnosticCodes: [ 2345 ]
}
[nodemon] app crashed - waiting for file changes before starting...

Comment (Scott-Kenning) [reply apps/backend/review/src/index.ts:17]: you can handle this error by adding a check below
`if(DB_ENDPOINT === "") {`
`    // handle error`
`}`

**Review by Scott-Kenning — APPROVED** [review]
Nothing blocking this from merging, but in the future the SQL returns should have only the fields required, and we should perform more validation on the inputs

File: apps/backend/review/src/createReview.ts (hunk 3 above)
Comment (MNThomson) [reply apps/backend/review/src/createReview.ts:17]: Server side validation for every property is required, frontend validation doesn't stop someone from making a network request
```console
curl -X POST -H "Content-Type: application/json" -d '{"stars": 1000, "comment": "", "listingId": 1}' http://localhost/api/review
```

File: apps/backend/review/src/index.ts (hunk 8 above)
Comment (MNThomson) [reply apps/backend/review/src/index.ts:17]:
```suggestion
const DB_ENDPOINT = process.env.DB_ENDPOINT;
if (!DB_ENDPOINT) {
	log + error here
}
```

File: apps/backend/review/src/createReview.ts (hunk 3 above)
Comment (bmachado18) [reply apps/backend/review/src/createReview.ts:17]: gotchya, updated!

**Review by MNThomson — APPROVED** [review]
Cinema Clap

**Review by Scott-Kenning — APPROVED** [review]

**Pull request merged**